src/payment/models/buy-crypto/services/buy-crypto-dex.service.ts

```diff
@@ -0,0 +1,236 @@
+      );
+    }
+
+    const amounts = transaction.amounts.map((a) => this.dexClient.parseAmount(a));
+
+    const { amount } = amounts.find((a) => a.asset === outputAsset);
+
+    if (!amount) {
+      throw new Error(`Failed to get amount for TX: ${purchaseTxId} while trying to extract purchased liquidity`);
+    }
+
+    return amount;
+  }
+
+  private async processNewBatches(
+    newBatches: BuyCryptoBatch[],
+    securedBatches: BuyCryptoBatch[],
+    pendingBatches: BuyCryptoBatch[],
+  ): Promise<void> {
+    for (const batch of newBatches) {
+      try {
+        const liquidity = await this.checkLiquidity(batch, securedBatches, pendingBatches);
+
+        if (liquidity !== 0) {
+          batch.secure(liquidity);
+          await this.buyCryptoBatchRepo.save(batch);
+
+          console.info(`Secured liquidity for batch. Batch ID: ${batch.id}`);
+
+          continue;
+        }
+
+        await this.purchaseLiquidity(batch);
+      } catch (e) {
+        console.info(`Error in processing new batch. Batch ID: ${batch.id}`, e);
+      }
+    }
+  }
+
+  private async checkLiquidity(
+    batch: BuyCryptoBatch,
+    securedBatches: BuyCryptoBatch[],
+    pendingBatches: BuyCryptoBatch[],
+  ): Promise<number> {
+    const securedAmount = securedBatches
+      .filter((securedBatch) => securedBatch.outputAsset === batch.outputAsset)
+      .reduce((acc, curr) => acc + curr.outputReferenceAmount, 0);
+
+    const pendingAmount = pendingBatches
+      .filter((pendingBatch) => pendingBatch.outputAsset === batch.outputAsset)
+      .reduce((acc, curr) => acc + curr.outputReferenceAmount, 0);
+
+    try {
+      const requiredAmount = await this.dexClient.testCompositeSwap(
+        batch.outputReferenceAsset,
+        batch.outputAsset,
+        batch.outputReferenceAmount + securedAmount + pendingAmount,
+      );
+
+      const availableAmount = await this.getAvailableTokenAmount(batch);
+
+      return availableAmount >= requiredAmount ? requiredAmount : 0;
+    } catch (e) {
+      console.error(`Error in checking liquidity for a batch, ID: ${batch.id}`, e);
+      throw e;
+    }
+  }
+
+  private async getAvailableTokenAmount(batch: BuyCryptoBatch): Promise<number> {
+    const tokens = await this.dexClient.getToken();
+    const token = tokens.map((t) => this.dexClient.parseAmount(t.amount)).find((pt) => pt.asset === batch.outputAsset);
+
+    return token ? token.amount : 0;
+  }
+
+  private async purchaseLiquidity(batch: BuyCryptoBatch) {
+    let txId: string;
+
+    try {
+      const DFIAmount =
+        (await this.dexClient.testCompositeSwap(batch.outputReferenceAsset, 'DFI', 1)) * batch.outputReferenceAmount;
+
+      txId = await this.dexClient.compositeSwap(
+        Config.node.dexWalletAddress,
+        'DFI',
+        Config.node.dexWalletAddress,
+        batch.outputAsset,
+        // swapping a bit more output asset to cover commissions
+        DFIAmount + Util.round(DFIAmount * 0.001, 8),
```

Comment:
Swap amount should be exact for all non-reference assets. For reference assets it should be a few percent more.